### ISSUE TITLE
[no-floating-promises] assigning unresolved promises doesn't cause an error

### ISSUE THREAD

**Issue by ATheCoder** [typescript-eslint/typescript-eslint#3359]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
module.exports = {
  env: {
    browser: true,
    amd: true,
    node: true,
  },
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: ['tsconfig.json'],
  },
  plugins: ['@typescript-eslint'],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  rules: {
    '@typescript-eslint/no-floating-promises': [2],
    '@typescript-eslint/strict-boolean-expressions': [2],
    "require-await": "off",
    "@typescript-eslint/no-misused-promises": [2]
  }
};

```

```TS
const foo = async () => console.log('a')
foo()
const bar = foo()
```

**Expected Result**
Both the second and third lines should be linted.

**Actual Result**

only the second line gets linted

**Additional Info**

I think assigning unresolved promises doesn't trigger `no-floating-promises`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.1` |
| `@typescript-eslint/parser`        | `4.22.1` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.25.0` |
| `node`                             | `14.16.1` |

**Comment by bradzacher** [comment]
This is intentional.

A floating promise is a promise that is left dangling and is not awaitable. Eg the second line.

Assigning a promise to a variable is not a floating promise because you can later `await` that variable.

It's a common pattern to do something like
```ts
const prom = somethingAsync();
// Do some work
await prom();
```
